File: src/main/java/org/warp/picalculator/math/rules/FractionsRule11.java
package org.warp.picalculator.math.rules;

import org.warp.picalculator.Error;
import org.warp.picalculator.math.Function;
import org.warp.picalculator.math.functions.Division;
import org.warp.picalculator.math.functions.Expression;
import org.warp.picalculator.math.functions.Multiplication;

import it.unimi.dsi.fastutil.objects.ObjectArrayList;

/**
 * Fractions rule<br>
 * <b>a / (b / c) = (a * c) / b</b>
 * 
 * @author Andrea Cavalli
 *
 */
public class FractionsRule11 {

	public static boolean compare(Function f) throws InterruptedException {
		final Division fnc = (Division) f;
		Function a;
		Function c;
		Division div2;
		if (fnc.getParameter2() instanceof Division) {
			div2 = (Division) fnc.getParameter2();
		} else {
			return false;
		}
		a = fnc.getParameter1();
		c = div2.getParameter2();
		return new Multiplication(fnc.getMathContext(), a, c).isSimplified() == false;
	}

	public static ObjectArrayList<Function> execute(Function f) throws Error {
		final ObjectArrayList<Function> result = new ObjectArrayList<>();
		final Division fnc = (Division) f;
		Function a;
		Function b;
		Function c;

		final Division div2 = (Division) fnc.getParameter2();

		a = fnc.getParameter1();
		b = div2.getParameter1();
		c = div2.getParameter2();
		result.add(new Division(fnc.getMathContext(), new Multiplication(fnc.getMathContext(), new Expression(fnc.getMathContext(), a), new Expression(fnc.getMathContext(), c)), b));

		return result;
	}

}

File: src/main/resources/rules/FractionsRule11.js
//Imports


var Error = org.warp.picalculator.Error;
var Function = org.warp.picalculator.math.Function;
var Division = org.warp.picalculator.math.functions.Division;
var Multiplication = org.warp.picalculator.math.functions.Multiplication;

var ObjectArrayList = Java.type("it.unimi.dsi.fastutil.objects.ObjectArrayList");

/**
 * Fractions rule
 * a / (b / c) = (a * c) / b
 * 
 * @author Andrea Cavalli
 *
 */
var rule = {
	// Rule name
	getRuleName: function() {
		return "FractionsRule11";
	},
	// Rule type
	getRuleType: function() {
		return RuleType.EXPANSION;
	},
	/* Rule function
	   Returns:
	     - null if it's not executable on the function "f"
	     - An ObjectArrayList<Function> if it did something
	*/

	execute: function(f) {
		var isExecutable;
		if (ScriptUtils.instanceOf(f, Division.class)) {
			var fnc = f;
			var a;
			var c;
			var div2;
			if (ScriptUtils.instanceOf(fnc.getParameter2(), Division.class)) {
				div2 = fnc.getParameter2();
				a = fnc.getParameter1();
				c = div2.getParameter2();
				isExecutable = true;
			} else {
				isExecutable = false;
			}
		}
		if (isExecutable) {
			var result = new ObjectArrayList();
			var fnc = f;
			var a;
			var b;
			var c;
	
			var div2 = fnc.getParameter2();
	
			a = fnc.getParameter1();
			b = div2.getParameter1();
			c = div2.getParameter2();
			result.add(new Division(fnc.getMathContext(), new Multiplication(fnc.getMathContext(), a, c), b));
	
			return result;
		} else {
			return null;
		}
	}
}

//Add this rule to the list of rules
RulesManager.addRule(engine.getInterface(rule, Rule.class));